spec/models/grace_period_submission_rule_spec.rb

```diff
@@ -5,6 +5,41 @@
     expect(rule.save).to be_truthy
   end
 
+  context 'When the student did not submit any assignment files' do
+    let(:student) { create(:student, grace_credits: 7) }
+    let(:grouping) { create(:grouping_with_inviter) }
+    let(:submission) { create(:submission, is_empty: true) }
```

Comment:
My comment about these `create` calls is that they're actually creating a bunch of unrelated objects. For example, the submission that's created on this line doesn't belong to the grouping created on line 10, nor does it belong to the assignment created on line 13. The Submission factory (`factories/submissions.rb`) automatically creates a separate grouping (with a brand-new assignment as well). This doesn't affect what happens in the test, because `rule.apply_submission_rule(submission)` works even when the submission is for a different assignment than `rule`, but it's still kind of misleading.

In fact, the existing test cases also test `apply_submission_rule`. So rather than create a new block here, my recommendation would be to take the logic from the tests you've written and add them as new cases inside the block on line 43; this way you can take advantage of the objects already created there. You can create a new nested context that tests for the case when a grouping has no files submitted.